src/lapack/gpu/lacpy.cu

```diff
@@ -0,0 +1,136 @@
+
+using namespace dlaf::util::cuda_operators;
+
+struct LacpyParams {
+  static constexpr unsigned kernel_tile_size_rows = 64;
+  static constexpr unsigned kernel_tile_size_cols = 16;
+};
+
+template <class T>
+__device__ void copyAll(const unsigned m, const unsigned n, const T* a, const unsigned lda, T* b,
+                        const unsigned ldb) {
+  constexpr unsigned kernel_tile_size_rows = LacpyParams::kernel_tile_size_rows;
+  constexpr unsigned kernel_tile_size_cols = LacpyParams::kernel_tile_size_cols;
+
+  const unsigned i = blockIdx.x * kernel_tile_size_rows + threadIdx.x;
+  const unsigned j = blockIdx.y * kernel_tile_size_cols;
+
+  if (i >= m)
+    return;
+
+  const unsigned k_max = min(j + kernel_tile_size_cols, n);
+
+  for (unsigned k = j; k < k_max; ++k)
+    b[i + k * ldb] = a[i + k * lda];
+}
+
+template <bool (*comp)(unsigned, unsigned), class T>
+__device__ void copyDiag(const unsigned m, const unsigned n, const T* a, const unsigned lda, T* b,
+                         const unsigned ldb) {
+  constexpr unsigned kernel_tile_size_rows = LacpyParams::kernel_tile_size_rows;
+  constexpr unsigned kernel_tile_size_cols = LacpyParams::kernel_tile_size_cols;
+
+  const unsigned i = blockIdx.x * kernel_tile_size_rows + threadIdx.x;
+  const unsigned j = blockIdx.y * kernel_tile_size_cols;
+
+  if (i >= m)
+    return;
+
+  const unsigned k_max = min(j + kernel_tile_size_cols, n);
+
+  for (unsigned k = j; k < k_max; ++k)
+    if (i == k or (comp(i, k)))
+      b[i + k * ldb] = a[i + k * lda];
+}
+
+template <class T>
+__global__ void lacpy(cublasFillMode_t uplo, const unsigned m, const unsigned n, const T* a,
+                      const unsigned lda, T* b, const unsigned ldb) {
+  constexpr unsigned kernel_tile_size_rows = LacpyParams::kernel_tile_size_rows;
+  constexpr unsigned kernel_tile_size_cols = LacpyParams::kernel_tile_size_cols;
+
+  DLAF_GPU_ASSERT_HEAVY(kernel_tile_size_rows % kernel_tile_size_cols == 0);
+  DLAF_GPU_ASSERT_HEAVY(kernel_tile_size_rows == blockDim.x);
+  DLAF_GPU_ASSERT_HEAVY(1 == blockDim.y);
+  DLAF_GPU_ASSERT_HEAVY(1 == blockDim.z);
+  DLAF_GPU_ASSERT_HEAVY(gridDim.x == ceilDiv(m, kernel_tile_size_rows));
+  DLAF_GPU_ASSERT_HEAVY(gridDim.y == ceilDiv(n, kernel_tile_size_cols));
+  DLAF_GPU_ASSERT_HEAVY(1 == gridDim.z);
+
+  const unsigned i = blockIdx.x;
+  const unsigned j = blockIdx.y * kernel_tile_size_cols / kernel_tile_size_rows;
+
+  // Note: if (i == j) the kernel tile contains parts of the diagonal
+
+  switch (uplo) {
+    case CUBLAS_FILL_MODE_LOWER:
+      if (i == j)
+        copyDiag<dlaf::util::isLower>(m, n, a, lda, b, ldb);
+      else if (i > j)
+        copyAll(m, n, a, lda, b, ldb);
+      break;
+    case CUBLAS_FILL_MODE_UPPER:
+      if (i == j)
+        copyDiag<dlaf::util::isUpper>(m, n, a, lda, b, ldb);
+      else if (i < j)
+        copyAll(m, n, a, lda, b, ldb);
+      break;
+    case CUBLAS_FILL_MODE_FULL:
+      copyAll(m, n, a, lda, b, ldb);
+      break;
```

Comment:
This code branch is never used.

Reply:
yep, it was there just in case we wanted to do an iteration of benchmarks and compare it with `cudaMemcpy2DAsync`.

> (note: the kernel implemented supports also the general case, even if it is not used).

If you prefer to remove it (sidenote: I tested it), I understand it. Just let me know 😉

Reply:
> (sidenote: I tested it)
Currently is tested, but as no test executes it, it might break without noticing...

Reply:
See 0b32ac385beefe2aa5ec73bec7b8164561fc824a